util/chplenv/chpl_arch.py

```diff
@@ -588,29 +603,42 @@ def get(flag, map_to_compiler=False, get_lcd=False):
     compiler_val = chpl_compiler.get(flag)
     isprgenv = compiler_is_prgenv(compiler_val)
     if map_to_compiler and not isprgenv:
-        # Map flag/arch to compiler flag/argument
+        # Map arch to compiler argument
         # Don't do this for PrgEnv compiles since the compiler driver
         # handles specialization.
         version = get_compiler_version(compiler_val)
-        (flag, arch) = argument_map.find(arch, compiler_val, version)
-    elif arch and arch != 'none' and arch != 'unknown':
-        if is_known_arm(arch):
-            flag = 'cpu'
-        else:
+        arch = argument_map.find(arch, compiler_val, version)
+    if arch and arch != 'none' and arch != 'unknown':
+        native_arch = get_native_machine()
+        # x86 uses -march= where the others use -mcpu=
+        is_x86_32 = native_arch.startswith('i') and native_arch.endswith('86')
+        if (native_arch in ['amd64', 'x86_64']) or is_x86_32:
             flag = 'arch'
+        else:
+            flag = 'cpu'
     else:
         flag = 'none'
 
     return arch_tuple(flag or 'none', arch or 'unknown')
 
 
-# Returns the default machine.
-#  * if arch is native/none/unknown, return result of get_native_machine
-#  * otherwise returns the machine type corresponding to the selected arch
+# Returns the default machine.  The flag argument is 'host' or 'target'.
+#
+# Note that we cannot deduce the machine type corresponding to a given
+# architecture.  For example, knowing that we are on a sandybridge does
+# not tell us whether we are running in 32-bit mode (sometimes designated
+# as i686) or 64-bit mode (x86_64).  For some architectures, the same is
+# true of big-endian vs. little-endian mode as well.  Consequently, we have
+# insufficient information about the target.  Therefore, the machine type must
+# be the same for host and target, though the subarchitecture can be different.
+# We can cross compile from sandybridge to broadwell but not from
+# x86_64 to aarch64.
+#
+# Because of all this, the flag argument is unused and we return the same
+# answer for host or target.
```

Comment:
Hmm, this is interesting and brings up some questions.
 * Could we keep track of the general architecture families (e.g. x86 variants, Power, ARM) and check that these are harmonious with the specific CPU for specialization?
 * Don't we know 64-bit or 32-bit mode from the platform (e.g. linux32 vs linux64) ?
 * Is big endian vs little endian normally reported by `uname -m`? Or is it something we should consider to be part of the platform as well, for machine types where it can vary?

I don't necessarily need this PR to do those things but I'd like to have a plan for handling some of these things, since I'm likely to run into them soon (e.g. as part of issue  #11740). So if the answers to these questions are obvious to you - it will help if you can share them.

Reply:
> * Could we keep track of the general architecture families (e.g. x86 variants, Power, ARM) and check that these are harmonious with the specific CPU for specialization?

The answer to this one is yes, though it requires knowledge about a lot of processors to which we do not have access for testing.  What we might be able to do is provide full coverage for the ones we can test, and make an educated guess for others.